error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/CanonicalConstantFieldName.java

```diff
@@ -0,0 +1,180 @@
+    linkType = CUSTOM,
+    severity = WARNING,
+    tags = LIKELY_ERROR)
+public final class CanonicalConstantFieldName extends BugChecker implements ClassTreeMatcher {
+  private static final long serialVersionUID = 1L;
+  private static final Matcher<Tree> IS_CONSTANT =
+      allOf(hasModifier(Modifier.STATIC), hasModifier(Modifier.FINAL));
+  private static final Matcher<Tree> IS_PRIVATE = hasModifier(Modifier.PRIVATE);
+  private static final Pattern TO_SNAKE_CASE = Pattern.compile("([a-z])([A-Z])");
+  private static final ImmutableSet<String> DEFAULT_EXCLUDED_CONSTANT_FIELD_NAMES =
+      ImmutableSet.of("serialVersionUID");
+  private static final String EXCLUDED_CONSTANT_FIELD_NAMES =
+      "CanonicalConstantFieldName:ExcludedConstantFliedNames";
+  private static final String IS_INCLUDE_PUBLIC_CONSTANT_FIELDS =
+      "CanonicalConstantFieldName:IncludePublicConstantFields";
+  private final ImmutableList<String> optionalExcludedConstantFliedNames;
+  private final boolean includePublicConstantFieldNames;
+
+  /** Instantiates a default {@link CanonicalConstantFieldName} instance. */
+  public CanonicalConstantFieldName() {
+    this(ErrorProneFlags.empty());
+  }
+
+  /**
+   * Instantiates a customized {@link CanonicalConstantFieldName}.
+   *
+   * @param flags Any provided command line flags.
+   */
+  @Inject
+  CanonicalConstantFieldName(ErrorProneFlags flags) {
+    optionalExcludedConstantFliedNames = getCanonicalizedLoggerName(flags);
+    includePublicConstantFieldNames = isIncludePrivateConstantFieldNames(flags);
+  }
+
+  @Override
+  public Description matchClass(ClassTree tree, VisitorState state) {
+    ImmutableList<String> classVariables =
+        tree.getMembers().stream()
+            .filter(member -> member.getKind() == Kind.VARIABLE)
+            .map(VariableTree.class::cast)
+            .map(ASTHelpers::getSymbol)
+            .map(VarSymbol::getSimpleName)
+            .map(Name::toString)
+            .collect(toImmutableList());
+
+    if (classVariables.isEmpty()) {
+      return Description.NO_MATCH;
+    }
+
+    SuggestedFix.Builder fixBuilder = SuggestedFix.builder();
+    for (Tree member : tree.getMembers()) {
+      if (member.getKind() == Kind.VARIABLE) {
+        VariableTree variableTree = (VariableTree) member;
+        if (!(IS_CONSTANT.matches(variableTree, state)
+            && isFieldAccessModifierApplicable(variableTree, state))) {
+          return Description.NO_MATCH;
+        }
+
+        VarSymbol variableSymbol = ASTHelpers.getSymbol(variableTree);
+        if (variableSymbol == null) {
+          return Description.NO_MATCH;
+        }
+
+        String variableName = variableSymbol.getSimpleName().toString();
+        if (!isVariableUpperSnakeCase(variableName) && !isVariableNameExcluded(variableName)) {
+          String replacement = toUpperSnakeCase(variableName);
+
+          if (!classVariables.contains(replacement)) {
```

Comment:
IIUC, here we avoid renaming a variable if the target name is already declared in the exact same scope. Due to the global search and replace that we do, I suspect that any usage of the target name in this `CompilationUnit` could cause ambiguity, shadowing or confusion. So for this case, too, a `TreeScanner` may be employed.

Reply:
> would you like to have a crack at resolving the surviving mutants?

Yes, sir 🖖 

> IIUC, here we avoid renaming a variable if the target name is already declared in the exact same scope. Due to the global search and replace that we do, I suspect that any usage of the target name in this CompilationUnit could cause ambiguity, shadowing or confusion. So for this case, too, a TreeScanner may be employed.

Will look into this 👀 

Thanks.